File: engineering-docs/schema-pack/examples/validate_examples.py
#!/usr/bin/env python3

import json
import sys
from pathlib import Path

from jsonschema import Draft202012Validator, FormatChecker
from referencing import Registry, Resource


ROOT = Path(__file__).resolve().parent
SCHEMA_ROOT = ROOT.parent

SCHEMA_FILE_MAP = {
    "action": "action.schema.json",
    "policy-outcome": "policy-outcome.schema.json",
    "snapshot-record": "snapshot-record.schema.json",
    "execution-result": "execution-result.schema.json",
    "run-event": "run-event.schema.json",
    "recovery-plan": "recovery-plan.schema.json",
    "timeline-step": "timeline-step.schema.json",
}

TRACE_COLLECTIONS = {
    "actions": "action",
    "policy_outcomes": "policy-outcome",
    "snapshot_records": "snapshot-record",
    "execution_results": "execution-result",
    "run_events": "run-event",
    "recovery_plans": "recovery-plan",
    "timeline_steps": "timeline-step",
}


def load_json(path: Path):
    return json.loads(path.read_text())


def build_validators():
    schemas = {}
    registry = Registry()
    for path in sorted(SCHEMA_ROOT.glob("*.schema.json")):
        schema = load_json(path)
        schemas[path.name] = schema
        schema_id = schema.get("$id")
        if schema_id:
            registry = registry.with_resource(schema_id, Resource.from_contents(schema))
    validators = {}
    for schema_name, file_name in SCHEMA_FILE_MAP.items():
        schema = schemas[file_name]
        validators[schema_name] = Draft202012Validator(
            schema,
            registry=registry,
            format_checker=FormatChecker(),
        )
    return validators


def collect_errors(validator, instance):
    return sorted(validator.iter_errors(instance), key=lambda e: list(e.absolute_path))


def validate_grouped_examples(validators, subdir: str, expect_valid: bool):
    failures = []
    base = ROOT / subdir
    for path in sorted(base.glob("*.examples.json")):
        payload = load_json(path)
        schema_name = payload["schema"]
        validator = validators[schema_name]
        for case_name, instance in payload["cases"].items():
            errors = collect_errors(validator, instance)
            if expect_valid and errors:
                failures.append(
                    f"{path.name}:{case_name} expected valid but failed with {errors[0].message}"
                )
            if not expect_valid and not errors:
                failures.append(
                    f"{path.name}:{case_name} expected invalid but passed validation"
                )
    return failures


def check_trace_referential_integrity(trace_name: str, records: dict):
    failures = []

    actions = {item["action_id"]: item for item in records.get("actions", [])}
    policy_outcomes = {
        item["policy_outcome_id"]: item for item in records.get("policy_outcomes", [])
    }
    snapshots = {item["snapshot_id"]: item for item in records.get("snapshot_records", [])}
    execution_results = {
        item["execution_result_id"]: item for item in records.get("execution_results", [])
    }
    recovery_plans = {
        item["recovery_plan_id"]: item for item in records.get("recovery_plans", [])
    }
    run_events = records.get("run_events", [])
    timeline_steps = records.get("timeline_steps", [])

    run_ids = set()
    for action in actions.values():
        run_ids.add(action["run_id"])
    for event in run_events:
        run_ids.add(event["run_id"])
    for step in timeline_steps:
        run_ids.add(step["run_id"])

    if len(run_ids) > 1:
        failures.append(f"{trace_name}: multiple run_ids present: {sorted(run_ids)}")

    for obj in policy_outcomes.values():
        if obj["action_id"] not in actions:
            failures.append(
                f"{trace_name}: policy outcome {obj['policy_outcome_id']} references missing action {obj['action_id']}"
            )

    for obj in snapshots.values():
        if obj["action_id"] not in actions:
            failures.append(
                f"{trace_name}: snapshot {obj['snapshot_id']} references missing action {obj['action_id']}"
            )

    for obj in execution_results.values():
        if obj["action_id"] not in actions:
            failures.append(
                f"{trace_name}: execution result {obj['execution_result_id']} references missing action {obj['action_id']}"
            )

    for obj in recovery_plans.values():
        target = obj["target"]
        action_id = target.get("action_id")
        snapshot_id = target.get("snapshot_id")
        if action_id and action_id not in actions:
            failures.append(
                f"{trace_name}: recovery plan {obj['recovery_plan_id']} references missing action {action_id}"
            )
        if snapshot_id and snapshot_id not in snapshots:
            failures.append(
                f"{trace_name}: recovery plan {obj['recovery_plan_id']} references missing snapshot {snapshot_id}"
            )

    for event in run_events:
        causality = event.get("causality", {})
        refs = event.get("refs", [])
        action_id = causality.get("action_id")
        policy_outcome_id = causality.get("policy_outcome_id")
        snapshot_id = causality.get("snapshot_id")
        execution_result_id = causality.get("execution_result_id")
        recovery_plan_id = causality.get("recovery_plan_id")

        if action_id and action_id not in actions:
            failures.append(f"{trace_name}: event {event['event_id']} references missing action {action_id}")
        if policy_outcome_id and policy_outcome_id not in policy_outcomes:
            failures.append(
                f"{trace_name}: event {event['event_id']} references missing policy outcome {policy_outcome_id}"
            )
        if snapshot_id and snapshot_id not in snapshots:
            failures.append(
                f"{trace_name}: event {event['event_id']} references missing snapshot {snapshot_id}"
            )
        if execution_result_id and execution_result_id not in execution_results:
            failures.append(
                f"{trace_name}: event {event['event_id']} references missing execution result {execution_result_id}"
            )
        if recovery_plan_id and recovery_plan_id not in recovery_plans:
            failures.append(
                f"{trace_name}: event {event['event_id']} references missing recovery plan {recovery_plan_id}"
            )

        for ref in refs:
            ref_id = ref["id"]
            kind = ref["kind"]
            if kind == "action" and ref_id not in actions:
                failures.append(f"{trace_name}: event {event['event_id']} refs missing action {ref_id}")
            if kind == "policy_outcome" and ref_id not in policy_outcomes:
                failures.append(
                    f"{trace_name}: event {event['event_id']} refs missing policy outcome {ref_id}"
                )
            if kind == "snapshot" and ref_id not in snapshots:
                failures.append(f"{trace_name}: event {event['event_id']} refs missing snapshot {ref_id}")
            if kind == "execution_result" and ref_id not in execution_results:
                failures.append(
                    f"{trace_name}: event {event['event_id']} refs missing execution result {ref_id}"
                )

    for step in timeline_steps:
        action_id = step.get("action_id")
        related = step["related"]
        snapshot_id = related.get("snapshot_id")
        execution_result_id = related.get("execution_result_id")
        for recovery_plan_id in related.get("recovery_plan_ids", []):
            if recovery_plan_id not in recovery_plans:
                failures.append(
                    f"{trace_name}: timeline step {step['step_id']} references missing recovery plan {recovery_plan_id}"
                )
        if action_id is not None and action_id not in actions:
            failures.append(
                f"{trace_name}: timeline step {step['step_id']} references missing action {action_id}"
            )
        if snapshot_id is not None and snapshot_id not in snapshots:
            failures.append(
                f"{trace_name}: timeline step {step['step_id']} references missing snapshot {snapshot_id}"
            )
        if execution_result_id is not None and execution_result_id not in execution_results:
            failures.append(
                f"{trace_name}: timeline step {step['step_id']} references missing execution result {execution_result_id}"
            )

    return failures


def validate_traces(validators):
    failures = []
    base = ROOT / "traces"
    for path in sorted(base.glob("*.json")):
        payload = load_json(path)
        records = payload["records"]
        for collection_name, schema_name in TRACE_COLLECTIONS.items():
            validator = validators[schema_name]
            for index, instance in enumerate(records.get(collection_name, []), start=1):
                errors = collect_errors(validator, instance)
                if errors:
                    failures.append(
                        f"{path.name}:{collection_name}[{index}] failed schema validation: {errors[0].message}"
                    )
        failures.extend(check_trace_referential_integrity(path.name, records))
    return failures


def main():
    validators = build_validators()

    failures = []
    failures.extend(validate_grouped_examples(validators, "valid", expect_valid=True))
    failures.extend(validate_grouped_examples(validators, "invalid", expect_valid=False))
    failures.extend(validate_traces(validators))

    if failures:
      print("Schema example validation failed:")
      for failure in failures:
          print(f"- {failure}")
      return 1

    print("Schema example validation passed.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: engineering-docs/schema-pack/examples/validate_examples.mjs
#!/usr/bin/env node

import fs from "node:fs/promises";
import { createRequire } from "node:module";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.dirname(fileURLToPath(import.meta.url));
const SCHEMA_ROOT = path.resolve(ROOT, "..");
const requireFromSchemasPackage = createRequire(new URL("../../../packages/schemas/package.json", import.meta.url));
const Ajv2020 = requireFromSchemasPackage("ajv/dist/2020").default;
const addFormats = requireFromSchemasPackage("ajv-formats");

const SCHEMA_FILE_MAP = {
  action: "action.schema.json",
  "policy-outcome": "policy-outcome.schema.json",
  "snapshot-record": "snapshot-record.schema.json",
  "execution-result": "execution-result.schema.json",
  "run-event": "run-event.schema.json",
  "recovery-plan": "recovery-plan.schema.json",
  "timeline-step": "timeline-step.schema.json",
};

const TRACE_COLLECTIONS = {
  actions: "action",
  policy_outcomes: "policy-outcome",
  snapshot_records: "snapshot-record",
  execution_results: "execution-result",
  run_events: "run-event",
  recovery_plans: "recovery-plan",
  timeline_steps: "timeline-step",
};

async function loadJson(filePath) {
  return JSON.parse(await fs.readFile(filePath, "utf8"));
}

async function listFiles(dirPath, predicate) {
  const entries = await fs.readdir(dirPath, { withFileTypes: true });
  return entries
    .filter((entry) => entry.isFile() && predicate(entry.name))
    .map((entry) => path.join(dirPath, entry.name))
    .sort();
}

async function buildValidators() {
  const ajv = new Ajv2020({ allErrors: true, strict: false });
  addFormats(ajv);

  const schemas = new Map();
  const schemaPaths = await listFiles(SCHEMA_ROOT, (name) => name.endsWith(".schema.json"));
  for (const schemaPath of schemaPaths) {
    const schema = await loadJson(schemaPath);
    schemas.set(path.basename(schemaPath), schema);
    ajv.addSchema(schema, schema.$id ?? path.basename(schemaPath));
  }

  const validators = new Map();
  for (const [schemaName, fileName] of Object.entries(SCHEMA_FILE_MAP)) {
    const schema = schemas.get(fileName);
    if (!schema) {
      throw new Error(`Missing schema file ${fileName}`);
    }
    const validator = ajv.getSchema(schema.$id) ?? ajv.compile(schema);
    validators.set(schemaName, validator);
  }

  return validators;
}

function collectErrors(validator, instance) {
  const valid = validator(instance);
  if (valid) {
    return [];
  }
  return [...(validator.errors ?? [])].sort((a, b) => String(a.instancePath).localeCompare(String(b.instancePath)));
}

function errorMessage(error) {
  return error.message ?? "unknown validation error";
}

async function validateGroupedExamples(validators, subdir, expectValid) {
  const failures = [];
  const base = path.join(ROOT, subdir);
  const examplePaths = await listFiles(base, (name) => name.endsWith(".examples.json"));

  for (const examplePath of examplePaths) {
    const payload = await loadJson(examplePath);
    const validator = validators.get(payload.schema);
    if (!validator) {
      failures.push(`${path.basename(examplePath)} references unknown schema ${payload.schema}`);
      continue;
    }

    for (const [caseName, instance] of Object.entries(payload.cases)) {
      const errors = collectErrors(validator, instance);
      if (expectValid && errors.length > 0) {
        failures.push(
          `${path.basename(examplePath)}:${caseName} expected valid but failed with ${errorMessage(errors[0])}`,
        );
      }
      if (!expectValid && errors.length === 0) {
        failures.push(`${path.basename(examplePath)}:${caseName} expected invalid but passed validation`);
      }
    }
  }

  return failures;
}

function checkTraceReferentialIntegrity(traceName, records) {
  const failures = [];

  const actions = new Map((records.actions ?? []).map((item) => [item.action_id, item]));
  const policyOutcomes = new Map((records.policy_outcomes ?? []).map((item) => [item.policy_outcome_id, item]));
  const snapshots = new Map((records.snapshot_records ?? []).map((item) => [item.snapshot_id, item]));
  const executionResults = new Map(
    (records.execution_results ?? []).map((item) => [item.execution_result_id, item]),
  );
  const recoveryPlans = new Map((records.recovery_plans ?? []).map((item) => [item.recovery_plan_id, item]));
  const runEvents = records.run_events ?? [];
  const timelineSteps = records.timeline_steps ?? [];

  const runIds = new Set();
  for (const action of actions.values()) {
    runIds.add(action.run_id);
  }
  for (const event of runEvents) {
    runIds.add(event.run_id);
  }
  for (const step of timelineSteps) {
    runIds.add(step.run_id);
  }

  if (runIds.size > 1) {
    failures.push(`${traceName}: multiple run_ids present: ${JSON.stringify([...runIds].sort())}`);
  }

  for (const obj of policyOutcomes.values()) {
    if (!actions.has(obj.action_id)) {
      failures.push(`${traceName}: policy outcome ${obj.policy_outcome_id} references missing action ${obj.action_id}`);
    }
  }

  for (const obj of snapshots.values()) {
    if (!actions.has(obj.action_id)) {
      failures.push(`${traceName}: snapshot ${obj.snapshot_id} references missing action ${obj.action_id}`);
    }
  }

  for (const obj of executionResults.values()) {
    if (!actions.has(obj.action_id)) {
      failures.push(`${traceName}: execution result ${obj.execution_result_id} references missing action ${obj.action_id}`);
    }
  }

  for (const obj of recoveryPlans.values()) {
    const actionId = obj.target?.action_id;
    const snapshotId = obj.target?.snapshot_id;
    if (actionId && !actions.has(actionId)) {
      failures.push(`${traceName}: recovery plan ${obj.recovery_plan_id} references missing action ${actionId}`);
    }
    if (snapshotId && !snapshots.has(snapshotId)) {
      failures.push(`${traceName}: recovery plan ${obj.recovery_plan_id} references missing snapshot ${snapshotId}`);
    }
  }

  for (const event of runEvents) {
    const causality = event.causality ?? {};
    const refs = event.refs ?? [];
    const actionId = causality.action_id;
    const policyOutcomeId = causality.policy_outcome_id;
    const snapshotId = causality.snapshot_id;
    const executionResultId = causality.execution_result_id;
    const recoveryPlanId = causality.recovery_plan_id;

    if (actionId && !actions.has(actionId)) {
      failures.push(`${traceName}: event ${event.event_id} references missing action ${actionId}`);
    }
    if (policyOutcomeId && !policyOutcomes.has(policyOutcomeId)) {
      failures.push(`${traceName}: event ${event.event_id} references missing policy outcome ${policyOutcomeId}`);
    }
    if (snapshotId && !snapshots.has(snapshotId)) {
      failures.push(`${traceName}: event ${event.event_id} references missing snapshot ${snapshotId}`);
    }
    if (executionResultId && !executionResults.has(executionResultId)) {
      failures.push(`${traceName}: event ${event.event_id} references missing execution result ${executionResultId}`);
    }
    if (recoveryPlanId && !recoveryPlans.has(recoveryPlanId)) {
      failures.push(`${traceName}: event ${event.event_id} references missing recovery plan ${recoveryPlanId}`);
    }

    for (const ref of refs) {
      if (ref.kind === "action" && !actions.has(ref.id)) {
        failures.push(`${traceName}: event ${event.event_id} refs missing action ${ref.id}`);
      }
      if (ref.kind === "policy_outcome" && !policyOutcomes.has(ref.id)) {
        failures.push(`${traceName}: event ${event.event_id} refs missing policy outcome ${ref.id}`);
      }
      if (ref.kind === "snapshot" && !snapshots.has(ref.id)) {
        failures.push(`${traceName}: event ${event.event_id} refs missing snapshot ${ref.id}`);
      }
      if (ref.kind === "execution_result" && !executionResults.has(ref.id)) {
        failures.push(`${traceName}: event ${event.event_id} refs missing execution result ${ref.id}`);
      }
    }
  }

  for (const step of timelineSteps) {
    const related = step.related;
    const snapshotId = related?.snapshot_id;
    const executionResultId = related?.execution_result_id;
    for (const recoveryPlanId of related?.recovery_plan_ids ?? []) {
      if (!recoveryPlans.has(recoveryPlanId)) {
        failures.push(`${traceName}: timeline step ${step.step_id} references missing recovery plan ${recoveryPlanId}`);
      }
    }
    if (step.action_id !== undefined && step.action_id !== null && !actions.has(step.action_id)) {
      failures.push(`${traceName}: timeline step ${step.step_id} references missing action ${step.action_id}`);
    }
    if (snapshotId !== undefined && snapshotId !== null && !snapshots.has(snapshotId)) {
      failures.push(`${traceName}: timeline step ${step.step_id} references missing snapshot ${snapshotId}`);
    }
    if (executionResultId !== undefined && executionResultId !== null && !executionResults.has(executionResultId)) {
      failures.push(`${traceName}: timeline step ${step.step_id} references missing execution result ${executionResultId}`);
    }
  }

  return failures;
}

async function validateTraces(validators) {
  const failures = [];
  const base = path.join(ROOT, "traces");
  const tracePaths = await listFiles(base, (name) => name.endsWith(".json"));

  for (const tracePath of tracePaths) {
    const payload = await loadJson(tracePath);
    const records = payload.records;

    for (const [collectionName, schemaName] of Object.entries(TRACE_COLLECTIONS)) {
      const validator = validators.get(schemaName);
      for (const [index, instance] of (records[collectionName] ?? []).entries()) {
        const errors = collectErrors(validator, instance);
        if (errors.length > 0) {
          failures.push(
            `${path.basename(tracePath)}:${collectionName}[${index + 1}] failed schema validation: ${errorMessage(
              errors[0],
            )}`,
          );
        }
      }
    }

    failures.push(...checkTraceReferentialIntegrity(path.basename(tracePath), records));
  }

  return failures;
}

async function main() {
  const validators = await buildValidators();
  const failures = [
    ...(await validateGroupedExamples(validators, "valid", true)),
    ...(await validateGroupedExamples(validators, "invalid", false)),
    ...(await validateTraces(validators)),
  ];

  if (failures.length > 0) {
    console.log("Schema example validation failed:");
    for (const failure of failures) {
      console.log(`- ${failure}`);
    }
    return 1;
  }

  console.log("Schema example validation passed.");
  return 0;
}

process.exitCode = await main();
